sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -81,6 +74,42 @@ trait StateStore {
     iterator()
   }
 
+  /**
+   * Return an iterator containing all the key-value pairs in the StateStore. Implementations must
+   * ensure that updates (puts, removes) can be made while iterating over this iterator.
+   */
+  def iterator(): Iterator[UnsafeRowPair]
+
+  /**
+   * Clean up the resource.
+   *
+   * The method name is to respect backward compatibility on [[StateStore]].
+   */
+  def abort(): Unit
+}
+
+/**
+ * Base trait for a versioned key-value store. Each instance of a `StateStore` represents a specific
+ * version of state data, and such instances are created through a [[StateStoreProvider]].
+ *
+ * Unlike [[ReadOnlyStateStore]], `abort` method may not be called if the `commit` method succeeds
+ * to commit the change. (`hasCommitted` returns `true`.) Otherwise, `abort` method will be called.
+ * Implementation should deal with resource cleanup in both methods, but also need to guard with
+ * double resource cleanup.
+ */
+trait StateStore extends ReadOnlyStateStore {
```

Comment:
Is this what we want? So all `StateStore` are `ReadOnlyStateStore`?

Reply:
Yes, because existing implementations only have StateStore and we have to support backward compatibility. Please share the idea if you have better one.

I think this change is clearer than initial version of the PR - we now enforce ReadOnlyStateStore that write is simply impossible in the interface level. (Previously I added test to verify any writes throw exception, and I simply removed the test as it is simply not possible now.)

Reply:
So if we need to check if an instance of state store is read-only, we can only check if it is `StateStore` but not `ReadOnlyStateStore`? I agree that enforcing read-only behavior at interface level sounds good. Just it looks a bit weird that an instance of `StateStore` is also `ReadOnlyStateStore`.

Maybe `ReadStateStore` or `StateStoreForRead` instead of `ReadOnlyStateStore`? Just suggestion.

Reply:
Ah OK the interface name looks to bring confusion. `ReadStateStore` looks OK to me. Thanks!